explanation/performance/perf-pgo.md

```diff
@@ -0,0 +1,153 @@
+
+```bash
+The 'numbers' are in 1000s of bytes per second processed.
+type             16 bytes     64 bytes    256 bytes   1024 bytes   8192 bytes  16384 bytes
+sha512           91713.74k   366632.96k   684349.30k  1060512.22k  1260359.00k  1273277.92k
+aes-256-cbc    1266018.36k  1437870.33k  1442743.14k  1449933.84k  1453336.84k  1453484.99k
+md5             140978.58k   367562.53k   714094.14k   955267.21k  1060969.81k  1068296.87k
+                  sign    verify    sign/s verify/s
+rsa 2048 bits 0.000173s 0.000012s   5776.8  86791.9
+```
+
+### Building OpenSSL for profiling
+
+Before we are able to profile OpenSSL, we need sure that we compile the software in a way that the generated program meets the requirements.  In our case, you can use the `file` command after building the binary to make sure that it has not been stripped:
+
+```bash
+$ file /usr/bin/openssl
+/usr/bin/openssl: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=42a5ac797b6981bd45a9ece3f67646edded56bdb, for GNU/Linux 3.2.0, with debug_info, not stripped
+```
+
+You can use the `eu-readelf` command (from the `elfutils` package) to determine whether any compression was performed with `dwz`:
+
+```bash
+$ eu-readelf -S /usr/bin/openssl
+[...]
+Section Headers:
+[Nr] Name                 Type         Addr             Off      Size     ES Flags Lk Inf Al
+[...]
+[31] .debug_abbrev        PROGBITS     0000000000000000 001bcbf3 000110c4  0        0   0  1
+[...]
+```
+
+If the `Flags` field has `C` in it, this means that `dwz` was used.  You will then need to figure out how to disable its execution during the build.
+
+### Using `perf` to obtain profile data
+
+With OpenSSL prepared and ready to be profiled, it's now time to use `perf` to obtain profiling data from our workload.  If you are on an Intel system, you can invoke `perf` using:
+
+```bash
+$ sudo perf record \
+	-e br_inst_retired.near_taken \
+	--branch-any \
+	--all-cpus \
+	--output openssl-nonpgo.perfdata \
+	-- openssl speed -mr -seconds 60 -evp md5 sha512 rsa2048 aes-256-cbc
+```
+
+On an AMD system, use the following invocation:
+
+```bash
+$ sudo perf record \
+	-e 'cpu/event=0xc4,umask=0x0,name=ex_ret_brn_tkn/' \
+	--branch-any \
+	--all-cpus \
+	--output openssl-nonpgo.perfdata \
+	-- openssl speed -mr -seconds 60 -evp md5 sha512 rsa2048 aes-256-cbc
+```
+
+Note that the only thing that differs is the PMU event to be monitored.
+
+After the command has finished running, you should see a file named `openssl-nonpgo.perfdata` in your current directory.  This is the file we will now need to convert to `gcov`.
```

Comment:
You have experimented with long profiling, long profiling at lower sampling frequency to keep size under control and more. Some people might not have a nice 60 second test. Would it be worth to elude to the existence of frequency control and present some file size estimates?

Reply:
Good idea.  For the file size estimates, I couldn't find my previous runs (I guess they were on horsea...), so I had to dig my notes/memory here.  I wasn't able to find a concrete value, so I ended up putting "several gigabytes" (which is true).  Hopefully that's good enough.